libacm/adecode.cpp

```diff
@@ -0,0 +1,161 @@
+  ACMStream *m_acm = nullptr;
+  ReadDataFunction m_readerFunction;
+  void *m_pReaderData;
+};
+
+/**************************************************************/
+/*                         Construction                       */
+/**************************************************************/
+
+int AcmReadFunc(void *ptr, int size, int n, void *datasrc) {
+  InternalAudioDecoder *iad = reinterpret_cast<InternalAudioDecoder *>(datasrc);
+  int ret =
+      iad->m_readerFunction(iad->m_pReaderData, ptr, (unsigned int)size * n);
+  // ret < 0: error, ret == 0: EOF, ret > 0: read ret bytes of data
+  // apparently acm_io_callbacks::read() expects pretty much the same behavior,
+  // except that for > 0 it's not number of bytes but number of items (like in
+  // fread())
+  if (ret > 0) {
+    ret /= size;
+  }
+  return ret;
+}
+
+InternalAudioDecoder::InternalAudioDecoder(ReadDataFunction readerFunction,
+                                           void *pReaderData)
+    : m_readerFunction(readerFunction), m_pReaderData(pReaderData) {}
+
+// Initialize the decoder
+bool InternalAudioDecoder::Initialize() {
+
+  acm_io_callbacks io = {
+      AcmReadFunc}; // set the read function, the others are optional
+
+  int force_channels =
+      0; // 0 = let libacm figure out how many channels the file has
+  // TODO: the old libacm.cpp was more optimistic about the numbers of channel
+  // from the file header
+  //       than libacm's decode.c, which assumes that channels are always >= 2
+  //       (unless it's WAVC instead of "plain ACM"), i.e. that a file header
+  //       specifying 1 is wrong. If it turns out that we really have ACM files
+  //       with just one channel (not unusual for ingame sounds?), we might have
+  //       to either patch acm_open_decoder() or somehow detect the number of
+  //       channels here and set force_channels accordingly
+  int ret = acm_open_decoder(&m_acm, this, io, force_channels);
+  return ret == ACM_OK;
+}
+
+/**************************************************************/
+/*                         Destruction                        */
+/**************************************************************/
+
+InternalAudioDecoder::~InternalAudioDecoder() {
+  if (m_acm != nullptr)
+    acm_close(m_acm);
+}
+
+/**************************************************************/
+/*                           Reading                          */
+/**************************************************************/
+
+// Read data from the audio decoder.
+//   pBuffer: The buffer to receive the data into
+//    amount: How much data to read
+// Returns the number of bytes read - zero when we're at the end of the file
+uint32 InternalAudioDecoder::Read(void *pBuffer, uint32 amount) {
+  const int bigendianp = 0; // we want little endian samples
+  const int wordlen = 2;    // the only supported value
+  const int sgned = 1;      // we want signed samples
+  uint32 totalBytesRead = 0;
+
+  while (totalBytesRead < amount) {
```

Comment:
This code appears to read over and over into the same place in the buffer, rather than appending to the buffer with each read.

Reply:
oh right - damn =)

(this is what I meant with sound on linux being generally broken, I couldn't tell if I did this right or not, just that it didn't crash anymore)